framework/framework.go

```diff
@@ -142,3 +142,60 @@ func (f *Framework) ProcessDelete(obj interface{}, resources []Resource) error {
 
 	return nil
 }
+
+// ProcessUpdate is a drop-in for an informer's UpdateFunc. It receives the
+// custom object observed during TPR watches and anything that implements
+// Resource. ProcessUpdate takes care about all necessary reconciliation logic
+// for update events. For complex resources this means state has to be created,
+// deleted and updated eventually, in this order.
+//
+//     func updateFunc(oldObj, newObj interface{}) {
+//         err := f.ProcessUpdate(oldObj, newObj, resources)
+//         if err != nil {
+//             // error handling here
+//         }
+//     }
+//
+//     newResourceEventHandler := &cache.ResourceEventHandlerFuncs{
+//         UpdateFunc:    updateFunc,
+//     }
+//
+func (f *Framework) ProcessUpdate(oldObj, newObj interface{}, resources []Resource) error {
```

Comment:
>  Should the old custom object just be removed from ProcessUpdate?

I believe this is the right approach. I even think that `ProcessCreate` shouldn't exist. Because they should have exactly the same semantics.

> The deployment has a name which is made up using information in the TPO. When the the information in the TPO change that are used to compute the deployment name then relying on the new TPO fails as well.

This would be bad operator design IMO. The operator should be able to retrieve all the resources it created one way or another. If the deployment name is dynamic then it should have a prefix or lables allowing to distinguish it from the deployments associated with different TPOs. If all of that is impossible the state should be tracked in separate database/storage.

We can't rely on previous TPO carnation. Another good example below.

1. Operator crash looping.
2. User makes a request to update cluster.
3. Operator fixed and starts.

In that case it's impossible to know previous value of the TPO. Our operators should be able to survive that.

Also when operator starts all TPOs are replayed as `OnAdd` events. Same happens after `ResyncPeriod`.

I can make more reasons here.